File: Cpp_src/medium/search_a_2D_matrix_II_240.cc
#include <iostream>
#include <vector>
using namespace std;

// Use search:
// Time Complexity: O(m+n)
// Space Complexity:O(1)
// Runtime: 56 ms, faster than 98.29%
// Memory Usage: 12.9 MB, less than 73.33%
class Solution
{
public:
    bool searchMatrix(vector<vector<int>> &matrix, int target)
    {
        if (matrix.size() == 0)
        {
            return false;
        }
        int size_row = matrix.size(), size_col = matrix[0].size();
        int row = 0, col = size_col - 1;

        while (row < size_row && col >= 0)
        {
            if (matrix[row][col] > target)
            {
                col--;
            }
            else if (matrix[row][col] < target)
            {
                row++;
            }
            else
            {
                return true;
            }
        }

        return false;
    }
};

File: Golang_src/medium/searchA2DMatrixII_240.go
package medium

// Use Search Space Reduction:
// m: row
// n: col
// Time Complexity: O(m+n)
// Space Complexity:O(l)
// Runtime: 34 ms, faster than 49.80%
// Memory Usage: 7.6 MB, less than 13.04%
func searchMatrixs(matrix [][]int, target int) bool {
	row, col := len(matrix), len(matrix[0])

	i, j := 0, col-1
	for i < row && j >= 0 {
		cur := matrix[i][j]

		if cur > target {
			j--
		} else if cur < target {
			i++
		} else {
			return true
		}
	}

	return false
}
